libs/utils/src/lsn.rs

```diff
@@ -13,10 +13,60 @@ use crate::seqwait::MonotonicCounter;
 pub const XLOG_BLCKSZ: u32 = 8192;
 
 /// A Postgres LSN (Log Sequence Number), also known as an XLogRecPtr
-#[derive(Clone, Copy, Eq, Ord, PartialEq, PartialOrd, Hash, Serialize, Deserialize)]
-#[serde(transparent)]
+#[derive(Clone, Copy, Eq, Ord, PartialEq, PartialOrd, Hash)]
 pub struct Lsn(pub u64);
 
+impl Serialize for Lsn {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        if serializer.is_human_readable() {
+            serializer.collect_str(self)
+        } else {
+            self.0.serialize(serializer)
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Lsn {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
```

Comment:
I found this PR from https://github.com/neondatabase/neon/issues/3511.
qq: Why wouldn't you use https://docs.rs/serde/latest/serde/trait.Deserializer.html#method.is_human_readable the same way as you did in Serializer?

Reply:
I at least do not understand this question, as the outer level does use it already...? The outer level will need to provide a "hint" about the type for bincode at least (deserialize_u64).

Reply:
Aah, you meant:

```
if d.is_human_readable() {
    let s: String = String::deserialize(d)?;
    // do the from_str parsing here?
} else {
    let arr = <[u8; 16]>::deserialize(d)?;
    Id(arr)
}
```

I guess I never thought about this as all of the documentation is about creating visitors. I think the result would be even more ackward error message in case something would go wrong, as the expecting would be from array or str..

Reply:
I think this approach is okay if we want less code addition in PR.
If we want good error message, then the custom visitor impl is better.
For now, I improved the `expecting` messages for custom visitor impls.https://github.com/neondatabase/neon/pull/5335/commits/990e252f08039e457c5036e0e0c624142779dace